File: templates/claude/hooks/abax-policy.py
#!/usr/bin/env python3
"""
Abax policy hook for Claude Code.

Mirrors the runtime enforcement that .opencode/plugins/abax-policy.ts
provides for opencode. Same merged policies file, different runtime.

Configured in .claude/settings.json under the `hooks` block:

  "hooks": {
    "PreToolUse":  [{ "matcher": "Task|Bash|Write|Edit",
                      "hooks": [{ "type": "command",
                                  "command": ".claude/hooks/abax-policy.py" }] }],
    "PostToolUse": [{ "matcher": "Task",
                      "hooks": [{ "type": "command",
                                  "command": ".claude/hooks/abax-policy.py" }] }]
  }

Three concerns, dispatched by hook_event_name + tool_name:

  PreToolUse  + Task                  → atomicity check
  PreToolUse  + Bash/Write/Edit/Task  → secret redaction
  PostToolUse + Task                  → runaway notice (never blocks)

Reads policies from .claude/policies/abax-policies.json (a verbatim copy
of the same file the opencode plugin reads — generator emits it to both
locations when the project targets claude).

Behaviour on missing/malformed policies file: fail-open (no-op). A
broken policy file should not block work.

Hook output protocol (Claude Code):
  - Exit 0 with no stdout = allow
  - Exit 0 with JSON stdout = structured response (decision, reason)
  - Exit 2 with stderr = block (Claude shows the stderr message)
  - Other exit codes = non-blocking error

We use exit 2 + stderr for block (clearest UX). For warn-level we
print to stderr but exit 0.
"""
import json
import os
import re
import sys
import time
from pathlib import Path

# ---- Inputs --------------------------------------------------------------

try:
    payload = json.load(sys.stdin)
except json.JSONDecodeError:
    # Malformed payload — let Claude proceed; nothing for us to enforce.
    sys.exit(0)

event = payload.get("hook_event_name", "")
tool_name = (payload.get("tool_name") or "").lower()  # case-insensitive
tool_input = payload.get("tool_input") or {}
session_id = payload.get("session_id", "")
cwd = payload.get("cwd") or os.getcwd()

# ---- Load policies (fail-open) -------------------------------------------

policies_path = Path(cwd) / ".claude" / "policies" / "abax-policies.json"
policies: dict = {}
if policies_path.is_file():
    try:
        policies = json.loads(policies_path.read_text())
    except Exception as e:
        sys.stderr.write(f"[abax-policy] could not parse {policies_path}: {e}\n")

# ---- Helpers -------------------------------------------------------------

def _block(reason: str) -> None:
    """Block the tool call — exit 2 with stderr."""
    sys.stderr.write(reason + "\n")
    sys.exit(2)

def _warn(reason: str) -> None:
    """Emit a warning, do not block."""
    sys.stderr.write(reason + "\n")

def _scannable_text(name: str, args: dict) -> list:
    """Pick text fields from tool_input that should be scanned for secrets."""
    if not isinstance(args, dict):
        return []
    if name == "bash":
        return [str(args.get("command", ""))]
    if name == "write":
        return [str(args.get("content", ""))]
    if name == "edit":
        return [str(args.get("new_string", "")), str(args.get("old_string", ""))]
    if name == "task":
        return [str(args.get("prompt", "")), str(args.get("description", ""))]
    # Generic fallback for custom tools
    return [str(v) for v in list(args.values())[:8] if isinstance(v, str)]

# ---- Concern 1: Secret redaction (PreToolUse on bash/write/edit/task) ----

if event == "PreToolUse":
    secret_patterns = (policies.get("secret_patterns") or {}).get("patterns") or []
    for text in _scannable_text(tool_name, tool_input):
        if not text:
            continue
        for p in secret_patterns:
            try:
                m = re.search(p["regex"], text)
            except re.error:
                continue
            if not m:
                continue
            sample = m.group(0)
            preview = f"{sample[:6]}…({len(sample)} chars)"
            sev = p.get("severity", "block")
            msg = (
                f"[abax-policy/secret] matched pattern \"{p['id']}\" "
                f"({p.get('description', '')}). Preview: {preview}."
            )
            if sev == "block":
                _block(msg + " Rotate the secret and retry.")
            else:
                _warn(msg)

# ---- Concern 2: Atomicity (PreToolUse on task) ---------------------------

if event == "PreToolUse" and tool_name == "task":
    contracts = policies.get("task_contracts") or {}
    actions_def = contracts.get("atomic_actions") or []
    combos = contracts.get("forbidden_combinations") or []
    exemptions = contracts.get("exemptions") or []

    prompt_text = str(tool_input.get("prompt") or tool_input.get("description") or "")
    role = str(
        tool_input.get("subagent_type")
        or tool_input.get("agent")
        or ""
    )
    if prompt_text and combos:
        lower = prompt_text.lower()
        detected = set()
        for a in actions_def:
            for kw in a.get("keywords", []):
                if kw.lower() in lower:
                    detected.add(a["id"])
                    break

        exempt_ids = set()
        for e in exemptions:
            if e.get("role") == role:
                exempt_ids.update(e.get("allow_combinations", []))

        violations = []
        for combo in combos:
            if combo["id"] in exempt_ids:
                continue
            if all(a in detected for a in combo["actions"]):
                violations.append(combo)

        if violations:
            summary = "\n".join(
                f"  • {v['id']}: {v['reason'].splitlines()[0]}" for v in violations
            )
            _block(
                f"[abax-policy/atomicity] blocked task delegation to "
                f"\"{role or '?'}\": forbidden combination(s) detected.\n"
                f"Detected actions: {', '.join(sorted(detected))}\n"
                f"Violations:\n{summary}\n"
                f"Split this Task into smaller atomic Tasks (one responsibility each)."
            )

# ---- Concern 3: Runaway notice (PostToolUse on task) ---------------------

# Track sub-session start times in /tmp keyed by tool_use_id. The hook
# script is short-lived (one process per call), so we persist via a file.
_RUNAWAY_DIR = Path("/tmp") / f"abax-runaway-{os.getuid()}"

if event == "PreToolUse" and tool_name == "task":
    _RUNAWAY_DIR.mkdir(exist_ok=True)
    use_id = payload.get("tool_use_id") or ""
    if use_id:
        (_RUNAWAY_DIR / use_id).write_text(str(time.time()))

if event == "PostToolUse" and tool_name == "task":
    use_id = payload.get("tool_use_id") or ""
    started = None
    if use_id:
        marker = _RUNAWAY_DIR / use_id
        if marker.is_file():
            try:
                started = float(marker.read_text())
            except Exception:
                started = None
            try:
                marker.unlink()
            except Exception:
                pass

    duration_min = (time.time() - started) / 60 if started else None

    # Output size as proxy for parts/tokens. Claude post-hook payload
    # carries `tool_response` with the result text.
    tool_response = payload.get("tool_response")
    if isinstance(tool_response, dict):
        # Some versions wrap it
        text = json.dumps(tool_response)
    else:
        text = str(tool_response or "")
    output_chars = len(text)
    PARTS_PROXY_FACTOR = 250
    parts_approx = output_chars // PARTS_PROXY_FACTOR

    role = str(
        tool_input.get("subagent_type")
        or tool_input.get("agent")
        or ""
    )

    limits_block = policies.get("runaway_limits") or {}
    role_categories = policies.get("role_categories") or {}
    cat = role_categories.get(role)
    by_role = (limits_block.get("by_role") or {}).get(role) or {}
    by_cat = (limits_block.get("by_category") or {}).get(cat or "") or {}
    default = limits_block.get("default") or {}

    def _resolve(field):
        return by_role.get(field) or by_cat.get(field) or default.get(field)

    parts_max = _resolve("parts_max")
    duration_max = _resolve("duration_min_max")

    breaches = []
    if parts_max and parts_approx > parts_max:
        breaches.append(f"parts≈{parts_approx} > {parts_max}")
    if duration_max and duration_min and duration_min > duration_max:
        breaches.append(f"duration={duration_min:.1f}min > {duration_max}")
    if breaches:
        _warn(
            f"[abax-policy/runaway] notice: task to \"{role or '?'}\" "
            f"exceeded limit(s): {'; '.join(breaches)}."
        )

# Default: allow
sys.exit(0)

File: templates/opencode/plugins/abax-policy.ts
/**
 * Abax policy plugin — generic enforcement of the manifest contract.
 *
 * Registered in opencode.json as:
 *   { "plugin": [".opencode/plugins/abax-policy.ts"] }
 *
 * Reads its baseline+overlay policies from `.opencode/policies/abax-policies.json`
 * (computed by the generator from data/rules/*.yaml + project-manifest.yaml).
 * Three concerns, one plugin:
 *
 *   1. Task atomicity — refuse `task` delegations whose prompt mixes
 *      forbidden combinations of atomic actions (e.g. fix + test +
 *      commit + push). Generic across roles; customisable per project.
 *
 *   2. Secret redaction — refuse / warn when bash, write, edit, or task
 *      args contain a credential matching one of the configured patterns
 *      (OpenAI keys, AWS keys, GitHub PATs, etc.).
 *
 *   3. Runaway detection — emit a notice when a `task` sub-session's
 *      output (size proxy for parts/tokens) exceeds the configured
 *      limit. Never blocks — the orchestrator decides what to do with
 *      the notice.
 *
 * Architecture decisions:
 *
 * - Single plugin file. opencode.json `plugin` is an array, but bundling
 *   three concerns in one file keeps the runtime lookup cheap and the
 *   import surface trivial.
 *
 * - Reads JSON, not YAML. The generator pre-merges baseline + overlay
 *   into JSON so the plugin has zero runtime dependencies.
 *
 * - All policies are silent no-ops if the JSON file is missing or any
 *   section is empty. Failing-open here is intentional: a broken policy
 *   file should not block work.
 *
 * - The plugin assumes `tool.execute.before` and `tool.execute.after`
 *   fire for the `task` tool. If your opencode build does not route
 *   task delegations through the tool layer, the atomicity and runaway
 *   checks become no-ops (secret redaction still works on bash/write/edit).
 *   See the bottom of this file for an empirical diagnostic logger you
 *   can enable by setting ABAX_POLICY_DEBUG=1.
 */
import type { Plugin } from "@opencode-ai/plugin";
import { readFileSync, existsSync, appendFileSync } from "node:fs";
import { join } from "node:path";

// ---- Policy types (mirror /srv/repos/Abax-Swarm/src/loader/schemas.ts) ----

interface AtomicAction {
  id: string;
  keywords: string[];
}
interface ForbiddenCombo {
  id: string;
  actions: string[];
  reason: string;
}
interface ContractExemption {
  role: string;
  allow_combinations: string[];
}
interface TaskContracts {
  atomic_actions: AtomicAction[];
  forbidden_combinations: ForbiddenCombo[];
  exemptions: ContractExemption[];
}

interface SecretPattern {
  id: string;
  regex: string;
  severity: "block" | "warn";
  description: string;
}
interface SecretPatterns {
  patterns: SecretPattern[];
}

interface Limits {
  parts_max?: number;
  duration_min_max?: number;
  tokens_max?: number;
}
interface RunawayLimits {
  default: Limits;
  by_category: Record<string, Limits>;
  by_role: Record<string, Limits>;
}

interface AbaxPolicies {
  task_contracts?: TaskContracts;
  secret_patterns?: SecretPatterns;
  runaway_limits?: RunawayLimits;
  /** Map roleId → category, used by runaway lookup. Generated alongside
   * the policies because the plugin doesn't know the team composition. */
  role_categories?: Record<string, string>;
}

// ---- Pure functions (module-private) ----
// NOTE: these are intentionally NOT exported. opencode's plugin loader
// (Bun-based) appears to introspect exported functions at load time — when
// they were exported, Bun emitted a misleading "X.toLowerCase is not a
// function" error during plugin init (the hooks still wired up correctly
// afterwards, but the noise made debugging harder). Keeping them private
// avoids the introspection path entirely.

/**
 * Detect which atomic actions appear in a free-text task prompt.
 * Case-insensitive substring match — keyword vocabularies are usually
 * short verbs/phrases so substring tolerates morphology ("compila",
 * "compilando"). Parameter named `text` (not `prompt`) to avoid the
 * Bun/Web `globalThis.prompt` global.
 */
function detectActions(text: string, actions: AtomicAction[]): Set<string> {
  const found = new Set<string>();
  const lower = text.toLowerCase();
  for (const a of actions) {
    for (const kw of a.keywords) {
      if (lower.includes(kw.toLowerCase())) {
        found.add(a.id);
        break;
      }
    }
  }
  return found;
}

/**
 * Find every forbidden_combination whose required actions are all detected.
 * Returns the matched combos so the caller can format the error.
 */
function findViolations(
  detected: Set<string>,
  combos: ForbiddenCombo[],
  role: string,
  exemptions: ContractExemption[],
): ForbiddenCombo[] {
  const exemptIds = new Set(
    exemptions.find((e) => e.role === role)?.allow_combinations ?? [],
  );
  const out: ForbiddenCombo[] = [];
  for (const combo of combos) {
    if (exemptIds.has(combo.id)) continue;
    if (combo.actions.every((a) => detected.has(a))) {
      out.push(combo);
    }
  }
  return out;
}

/**
 * Scan text for any secret pattern. Returns matches with their pattern
 * metadata. Caller decides whether to block (severity:block) or just warn.
 */
function scanSecrets(text: string, patterns: SecretPattern[]): Array<{
  pattern: SecretPattern;
  preview: string;
}> {
  const hits: Array<{ pattern: SecretPattern; preview: string }> = [];
  for (const p of patterns) {
    let re: RegExp;
    try {
      re = new RegExp(p.regex);
    } catch {
      continue;
    }
    const m = text.match(re);
    if (m) {
      // Redacted preview: first 6 chars + length, never the full match.
      const sample = m[0];
      const preview = `${sample.slice(0, 6)}…(${sample.length} chars)`;
      hits.push({ pattern: p, preview });
    }
  }
  return hits;
}

/**
 * Resolve effective limits for a role: by_role > by_category > default.
 * Returns the merged limits object (every field defined or undefined).
 */
function resolveLimits(
  role: string,
  limits: RunawayLimits | undefined,
  roleCategory: Record<string, string> | undefined,
): Limits {
  if (!limits) return {};
  const byRole = limits.by_role[role];
  const cat = roleCategory?.[role];
  const byCat = cat ? limits.by_category[cat] : undefined;
  const def = limits.default ?? {};
  return {
    parts_max: byRole?.parts_max ?? byCat?.parts_max ?? def.parts_max,
    duration_min_max:
      byRole?.duration_min_max ?? byCat?.duration_min_max ?? def.duration_min_max,
    tokens_max: byRole?.tokens_max ?? byCat?.tokens_max ?? def.tokens_max,
  };
}

// ---- Plugin entry point ----

const PLUGIN: Plugin = async (input) => {
  const policiesPath = join(input.directory, ".opencode/policies/abax-policies.json");
  let policies: AbaxPolicies = {};
  if (existsSync(policiesPath)) {
    try {
      policies = JSON.parse(readFileSync(policiesPath, "utf8"));
    } catch (err) {
      // Fail-open with a stderr breadcrumb. We never want a malformed policy
      // file to block work.
      console.error(`[abax-policy] could not parse ${policiesPath}: ${(err as Error).message}`);
    }
  }

  const debugLog = process.env.ABAX_POLICY_DEBUG ? join(input.directory, ".opencode/abax-policy-debug.log") : null;
  const dbg = (msg: string) => {
    if (debugLog) appendFileSync(debugLog, `${new Date().toISOString()} ${msg}\n`);
  };

  // Track sub-session start times for the runaway check.
  const taskStarts = new Map<string, number>();

  return {
    "tool.execute.before": async (i, o) => {
      const tool = (i.tool ?? "").toLowerCase();
      dbg(`before tool=${tool} args.keys=${Object.keys(o.args ?? {}).join(",")}`);

      // ---- Secret redaction (applies to any tool that carries text) ----
      if (policies.secret_patterns) {
        const fields = collectScannableText(tool, o.args);
        for (const text of fields) {
          const hits = scanSecrets(text, policies.secret_patterns.patterns);
          for (const h of hits) {
            if (h.pattern.severity === "block") {
              throw new Error(
                `[abax-policy/secret] blocked: matched pattern "${h.pattern.id}" (${h.pattern.description}). Preview: ${h.preview}. Rotate the secret and retry.`,
              );
            } else {
              console.warn(
                `[abax-policy/secret] WARN: matched pattern "${h.pattern.id}" (${h.pattern.description}) in tool=${tool}. Preview: ${h.preview}.`,
              );
            }
          }
        }
      }

      // ---- Atomicity check (only for task delegations) ----
      if (tool === "task" && policies.task_contracts) {
        // Local var named `taskPrompt` (not `prompt`) — see detectActions docstring.
        const taskPrompt = String(o.args?.prompt ?? o.args?.description ?? "");
        const role = String(o.args?.subagent_type ?? o.args?.agent ?? "");
        if (taskPrompt) {
          const actions = detectActions(taskPrompt, policies.task_contracts.atomic_actions);
          const violations = findViolations(
            actions,
            policies.task_contracts.forbidden_combinations,
            role,
            policies.task_contracts.exemptions,
          );
          if (violations.length > 0) {
            const summary = violations.map((v) => `  • ${v.id}: ${v.reason.split("\n")[0]}`).join("\n");
            throw new Error(
              `[abax-policy/atomicity] blocked task delegation to "${role || "?"}": forbidden combination(s) detected.\n` +
                `Detected actions: ${[...actions].join(", ")}\n` +
                `Violations:\n${summary}\n` +
                `Split this Task into smaller atomic Tasks (one responsibility each).`,
            );
          }
          taskStarts.set(i.callID, Date.now());
        }
      }
    },

    "tool.execute.after": async (i, o) => {
      const tool = (i.tool ?? "").toLowerCase();
      dbg(`after tool=${tool} output.size=${(o.output ?? "").length}`);

      // ---- Runaway notice (only for task) ----
      if (tool === "task" && policies.runaway_limits) {
        const role = String(i.args?.subagent_type ?? i.args?.agent ?? "");
        const limits = resolveLimits(role, policies.runaway_limits, policies.role_categories);
        const start = taskStarts.get(i.callID);
        const durationMin = start ? (Date.now() - start) / 60000 : undefined;
        taskStarts.delete(i.callID);

        // Heuristic: opencode doesn't expose parts count to the post-hook,
        // so we use output length as a proxy for "size of the sub-session
        // result". For a precise parts count, the orchestrator would need
        // to query the SDK; that path is documented but not implemented here.
        const outputChars = (o.output ?? "").length;
        const PARTS_PROXY_FACTOR = 250; // ~250 chars per part avg in practice
        const partsApprox = Math.round(outputChars / PARTS_PROXY_FACTOR);

        const breaches: string[] = [];
        if (limits.parts_max && partsApprox > limits.parts_max) {
          breaches.push(`parts≈${partsApprox} > ${limits.parts_max}`);
        }
        if (limits.duration_min_max && durationMin && durationMin > limits.duration_min_max) {
          breaches.push(`duration=${durationMin.toFixed(1)}min > ${limits.duration_min_max}`);
        }
        if (breaches.length > 0) {
          // NEVER block. Stderr notice — the orchestrator (or a separate
          // monitor) can pick this up and decide whether to escalate.
          console.warn(
            `[abax-policy/runaway] notice: task to "${role || "?"}" exceeded limit(s): ${breaches.join("; ")}.`,
          );
        }
      }
    },
  };
};

/**
 * Pick out the string fields of `args` that should be scanned for secrets.
 * Tool-specific because args shape varies.
 */
function collectScannableText(tool: string, args: any): string[] {
  if (!args || typeof args !== "object") return [];
  switch (tool) {
    case "bash":
      return [String(args.command ?? "")];
    case "write":
      return [String(args.content ?? "")];
    case "edit":
      return [String(args.new_string ?? ""), String(args.old_string ?? "")];
    case "task":
      return [String(args.prompt ?? ""), String(args.description ?? "")];
    default:
      // Generic fallback: scan every string-valued top-level arg. Catches
      // custom tools that carry text (e.g. send-message, http-post).
      return Object.values(args)
        .filter((v): v is string => typeof v === "string")
        .slice(0, 8); // cap to avoid pathological cases
  }
}

export default PLUGIN;
